preprocessors/ner/ner.py

```diff
@@ -0,0 +1,219 @@
+"""
+def find_index(text, word):
+    index = 1
+    for i in text.split():
+        if word in i:
+            return index
+        index += 1
+    return -1
+
+
+@app.route('/preprocessor', methods=['POST', 'GET'])
+def main():
+    
+    logging.debug("Received request")
+
+    with open('./schemas/preprocessors/ner.schema.json') as jsonfile:
+        data_schema = json.load(jsonfile)
+
+    with open('./schemas/preprocessor-response.schema.json') as jsonfile:
+        schema = json.load(jsonfile)
+
+    with open('./schemas/definitions.json') as jsonfile:
+        definition_schema = json.load(jsonfile)
+
+    with open('./schemas/request.schema.json') as jsonfile:
+        first_schema = json.load(jsonfile)
+
+    schema_store = {
+        data_schema['$id']: data_schema,
+        schema['$id']: schema,
+        definition_schema['$id']: definition_schema
+    }
+
+    resolver = jsonschema.RefResolver.from_schema(
+        schema, store=schema_store)
+
+    content = request.get_json()
+
+    try:
+        validator = jsonschema.Draft7Validator(first_schema, resolver=resolver)
+        validator.validate(content)
+    except jsonschema.exceptions.ValidationError as e:
+        logging.error(e)
+        return jsonify("Invalid Preprocessor JSON format"), 400
+
+    # ------ START COMPUTATION ------ #
+
+    name_ = "1"
+    captions = {}                            # dict to store the captions
+    text = get_alt(content)                  # extract alt text and save it in the caption dict
+    captions[name_] = text
+    html_ = content['graphic'] # get the image html content
+    
+    # save pic and caption to directory for clipscore evaluation
+    with open(CONTEXT_DIR+'/captions.json', "w") as outfile:
+        json.dump(captions, outfile)
+    save_image(html_, name_, IMAGE_DIR)
+
+    # check if we have an alt text
+    if len(text) < 2:
+        logging.info("No alttxt")
+        return "", 204
+
+    # create path parameters for clipscore
+    parameters = Namespace(candidates_json=CONTEXT_DIR+'/captions.json', compute_other_ref_metrics=1, image_dir=IMAGE_DIR, references_json=None, save_per_instance=CONTEXT_DIR+'/score.json')
+    
+    # calculate the clipscore
+    score = round(clipscore.main(parameters)['1']['CLIPScore'], 3)
+    # compute the NERs
+    ners = stanford_ner(text)
+    ner_data = []
+    for i in ners:
+        my_dict = {}
+        my_dict['value'] = i[0]
+        my_dict['tag'] = i[1]
+        my_dict['index'] = find_index(text, i[0])
+        ner_data.append(my_dict)
+
+    data = {
+        'clipscore': score,
+        'ner': ner_data,
+        'alttxt': captions['1']
+        }
+    
+    # ------ END COMPUTATION ------ #
+    
+    request_uuid = content["request_uuid"]
+    timestamp = time.time()
+    name = "ca.mcgill.a11y.image.preprocessor.ner"
+
+    response = {
+        'request_uuid': request_uuid,
+        'timestamp': int(timestamp),
+        'name': name,
+        'data': data
+    }
+
+    try:
```

Comment:
Good! Will eventually be necessary to check against the data schema for NER (that you are still working on).

Reply:
Not sure I understand this, I thought I'm doing it just below it:

```
    try:
        validator = jsonschema.Draft7Validator(schema, resolver=resolver)
        validator.validate(response)
    except jsonschema.exceptions.ValidationError as e:
        logging.error(e)
        return jsonify("Invalid Preprocessor JSON format"), 500

    logging.debug("Sending response")
    return response
```

Reply:
This only checks that you're returning something in the expected response format. It doesn't check that the content of the `data` key actually makes sense for the NER preprocessor `name`.